### PR TITLE
Lower hand

### PR DESCRIPTION
**Pull request by liumengyuan1997**
Dear reviewer,

I have implemented functionality to lower a participant's hand and lower all hands for the meeting moderator. There are also some remaining code sections related to creating notifications for the next speaker. I have commented out notification-related listeners as these features are not yet fully implemented. I look forward to receiving your comments and advice on these updates.

All the best,
Mengyuan

### CODE REVIEW

File: react/features/base/participants/middleware.ts [hunk 1]
```diff
@@ -236,10 +236,13 @@ MiddlewareRegistry.register(store => next => action => {
...
-            queue.push({
-                id: participant.id,
-                raisedHandTimestamp: participant.raisedHandTimestamp
-            });
+            // queue.push({
+            //     id: participant.id,
+            //     raisedHandTimestamp: participant.raisedHandTimestamp
+            // });
+
+            queue = [ ...queue, { id: participant.id,
```
Comment (saghul) [inline react/features/base/participants/middleware.ts:244]: Why is this change necessary?

File: lang/main.json [hunk 2]
```diff
@@ -786,6 +786,7 @@
         "newDeviceAction": "Use",
         "newDeviceAudioTitle": "New audio device detected",
         "newDeviceCameraTitle": "New camera detected",
+        "nextToSpeak": "Your turn to speak might come soon, get ready",
```
Comment (saghul) [inline lang/main.json:789]: Let's do the next to speak as a separate PR, since it's not related to lowering the hand.

File: react/features/base/participants/subscriber.ts [hunk 3]
```diff
@@ -1,8 +1,10 @@
+
```
Comment (saghul) [inline react/features/base/participants/subscriber.ts:1]: Please undo this.

File: react/features/participants-pane/components/web/FooterContextMenu.tsx [hunk 4]
```diff
@@ -147,6 +151,9 @@ export const FooterContextMenu = ({ isOpen, onDrawerClose, onMouseLeave }: IProp
                     onClick: muteAllVideo,
                     text: t('participantsPane.actions.stopEveryonesVideo')
                 } ] } />
+            {raisedHandsQueue.length !== 0 && (<LowerHandButton
```
Comment (saghul) [inline react/features/participants-pane/components/web/FooterContextMenu.tsx:154]: Is this check correct also when it's just the local user the one who raised the hand?

File: react/features/video-menu/actions.web.ts [hunk 5]
```diff
@@ -15,3 +17,20 @@ export function renderConnectionStatus(showConnectionInfo: boolean) {
         showConnectionInfo
     };
 }
+
+/**
+ * Notify the first participant in the speaker queue to prepare to speak.
+ *
+ * @returns {Function} A function that dispatches a notification action.
+ */
+export function notifyToSpeak() {
```
Comment (saghul) [inline react/features/video-menu/actions.web.ts:26]: Note for the next PR: don't put this in `actions.web` since we'll want to reuse it for mobile too.

File: react/features/video-menu/components/web/LowerHandButton.tsx [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    conference,
+    participantID
+}: IProps): JSX.Element => {
+    const { t } = useTranslation();
+    const localParticipant = useSelector(getLocalParticipant);
+    const _isModerator = Boolean(localParticipant?.role === PARTICIPANT_ROLE.MODERATOR);
+
+    const handleClick = useCallback(() => {
+        try {
+            _isModerator && conference?.sendEndpointMessage(
```
Comment (saghul) [inline react/features/video-menu/components/web/LowerHandButton.tsx:34]: This button should not be visible for non-moderators. Are we testing that?

File: react/features/conference/middleware.web.ts [hunk 7]
```diff
@@ -19,6 +20,8 @@ MiddlewareRegistry.register(store => next => action => {
                 /* onAllow */ () => APP.store.dispatch(setCameraFacingMode(data.facingMode)),
                 /* initiatorId */ participant.getId()
             ));
+        } else if (data?.name === LOWER_HAND) {
```
Comment (saghul) [inline react/features/conference/middleware.web.ts:23]: Check that the sender is a moderator.

**Review by saghul — COMMENTED** [review]
Great work so far! Left a few comments, let's discuss today.

File: react/features/video-menu/components/web/LowerHandButton.tsx (hunk 6 above)
Comment (liumengyuan1997) [reply react/features/video-menu/components/web/LowerHandButton.tsx]: Hi Saghul, I solved it by adding _isModerator condition in the return part. Please let me know if it is not recommended.

File: react/features/participants-pane/components/web/FooterContextMenu.tsx (hunk 4 above)
Comment (liumengyuan1997) [reply react/features/participants-pane/components/web/FooterContextMenu.tsx:154]: Hi Saghul, I fixed this part by adding a condition `if (!participantID) {
                dispatch(raiseHand(false));
            }` in the LowerHandButton.tsx.

File: react/features/base/tracks/constants.ts [hunk 8]
```diff
@@ -2,3 +2,4 @@
  * The payload name for remotely setting the camera facing mode message.
  */
 export const CAMERA_FACING_MODE_MESSAGE = 'camera-facing-mode-message';
+export const LOWER_HAND = 'lower-hand';
```
Comment (saghul) [inline react/features/base/tracks/constants.ts:5]: Let's call this LOWER_HAND_MESSAGE

File: react/features/conference/middleware.web.ts [hunk 9]
```diff
@@ -19,6 +20,8 @@ MiddlewareRegistry.register(store => next => action => {
                 /* onAllow */ () => APP.store.dispatch(setCameraFacingMode(data.facingMode)),
                 /* initiatorId */ participant.getId()
             ));
+        } else if (data?.name === LOWER_HAND && data?.isModerator) {
```
Comment (saghul) [inline react/features/conference/middleware.web.ts:23]: Don't check the payload, since that could be set to whatever, check that the sender (the `participant` variable) is a moderator.

File: react/features/participants-pane/components/web/FooterContextMenu.tsx [hunk 10]
```diff
@@ -147,6 +151,9 @@ export const FooterContextMenu = ({ isOpen, onDrawerClose, onMouseLeave }: IProp
...
+                conference = { conference }
+                participantID = '' />)}
```
Comment (saghul) [inline react/features/participants-pane/components/web/FooterContextMenu.tsx:156]: Please give this prop a default value so we don't pass it empty.

File: react/features/video-menu/components/web/LowerHandButton.tsx [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    const handleClick = useCallback(() => {
+        try {
+            if (!participantID) {
+                dispatch(raiseHand(false));
+            }
+            conference?.sendEndpointMessage(
+                participantID,
+                {
+                    name: LOWER_HAND,
+                    isModerator: _isModerator
```
Comment (saghul) [inline react/features/video-menu/components/web/LowerHandButton.tsx:43]: As mentioned above, let's not pass this here but rather check if the sender of the message is a moderator (on the receiving side).

File: react/features/video-menu/components/web/LowerHandButton.tsx [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+                }
+            );
+
+        } catch (error) {
```
Comment (saghul) [inline react/features/video-menu/components/web/LowerHandButton.tsx:47]: Drop this, none of the APIs there can throw an error.

File: react/features/video-menu/components/web/LowerHandButton.tsx [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+            logger.error('Error in lower hand:', error.message);
+        }
+    }, [ participantID ]);
+
+    return (
+        <React.Fragment>
+            {
+                _isModerator ? <ContextMenuItem
+                    accessibilityLabel = {
+                        participantID
```
Comment (saghul) [inline react/features/video-menu/components/web/LowerHandButton.tsx:57]: The label and text are the same, perhaps the content can be extracted to a variable and used in both places.

**Review by saghul — CHANGES_REQUESTED** [review]

**Comment by saghul** [comment]
@liumengyuan1997 Let us know when this is ready for another review round.

**Comment by liumengyuan1997** [comment]
> @liumengyuan1997 Let us know when this is ready for another review round.

Hi Saghul, this PR is ready for another review. Thanks! @saghul

File: react/features/video-menu/components/native/LowerHandButton.tsx [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+    icon = IconRaiseHand;
+    accessibilityLabel = 'participantsPane.actions.lowerHand';
+    label = 'participantsPane.actions.lowerHand';
+
+    /**
+     * Gets the current label.
+     *
+     * @returns {string}
+     */
+    _getLabel() {
```
Comment (saghul) [inline react/features/video-menu/components/native/LowerHandButton.tsx:40]: You can delete this.

File: react/features/video-menu/components/native/LowerHandButton.tsx [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+
+        return this.label;
+    }
+
+    /**
+     * Gets the current icon.
+     *
+     * @returns {string}
+     */
+    _getIcon() {
```
Comment (saghul) [inline react/features/video-menu/components/native/LowerHandButton.tsx:50]: You can drop this.

File: react/features/participants-pane/components/web/FooterContextMenu.tsx [hunk 16]
```diff
@@ -85,6 +88,8 @@ interface IProps {
 export const FooterContextMenu = ({ isOpen, onDrawerClose, onMouseLeave }: IProps) => {
     const dispatch = useDispatch();
     const isModerationSupported = useSelector((state: IReduxState) => isAvModerationSupported()(state));
+    const currentConference = useSelector(getCurrentConference);
```
Comment (saghul) [inline react/features/participants-pane/components/web/FooterContextMenu.tsx:91]: Rather than here, get it in mapStateToProps.

File: react/features/video-menu/components/native/LowerHandButton.tsx [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+
+/**
+ * Maps part of the Redux state to the props of this component.
+ *
+ * @param {Object} state - The Redux state.
+ * @param {Object} ownProps - Properties of component.
+ * @returns {IProps}
+ */
+function mapStateToProps(state: IReduxState, ownProps: any) {
+    const { participantID, conference } = ownProps;
```
Comment (saghul) [inline react/features/video-menu/components/native/LowerHandButton.tsx:80]: Rather than getting the conference as an own prop, get it ffrom the store here.

File: react/features/video-menu/components/native/RemoteVideoMenu.tsx [hunk 18]
```diff
@@ -256,8 +273,11 @@ function _mapStateToProps(state: IReduxState, ownProps: any) {
     const moderator = isLocalParticipantModerator(state);
     const _iAmVisitor = state['features/visitors'].iAmVisitor;
     const _isBreakoutRoom = isInBreakoutRoom(state);
+    const { conference } = state['features/base/conference'];
```
Comment (saghul) [inline react/features/video-menu/components/native/RemoteVideoMenu.tsx:276]: Ditto.

File: react/features/video-menu/components/web/LowerHandButton.tsx [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+/**
+ * Implements a React {@link Component} which displays a button for notifying certain
+ * participant who raised hand to lower hand.
+ *
+ * @returns {JSX.Element}
+ */
+const LowerHandButton = ({
+    currentConference,
```
Comment (saghul) [inline react/features/video-menu/components/web/LowerHandButton.tsx:31]: Use a selector to get the current conference.

**Comment by liumengyuan1997** [comment]
Hi Saghul, I finished editing the code. Please review it. Thanks!

**Review by saghul — APPROVED** [review]
LGTM! @mihhu PTAL and let's 🚀  this!

**Review by mihhu — APPROVED** [review]

**Comment by saghul** [comment]
Congratulations @liumengyuan1997 your PR just made it! 🚀
